# ImmregInformer: an interpretable Transformer-based method for prioritizing immune-regulatory cancer drivers
===========================================================================


[![license](https://img.shields.io/badge/python_-3.8.0_-blue)](https://www.python.org/)
[![license](https://img.shields.io/badge/torch_-1.12.0_-blue)](https://pytorch.org/)
[![license](https://img.shields.io/badge/scanpy_-1.9.0_-blue)](https://scanpy.readthedocs.io/en/stable/)
[![license](https://img.shields.io/badge/anndata_-0.8.0_-blue)](https://anndata-tutorials.readthedocs.io/en/latest/index.html/)
[![license](https://img.shields.io/badge/R_-4.2.2_-blue)](https://www.r-project.org/)

The identification and prioritization of immune-regulatory cancer driver mutations present a promising study for precision immunotherapy of cancer but remain considerable challenges. As there is no existing proven and efficient method to systematically explore the regulatory relationship between cancer driver mutations and immune response, we introduced a novel method, ImmregInformer, that leverages the powerful self-attention mechanisms of the Transformer model and the lasso-regularised ordinal regression to address these challenges. In particular, our approach integrated the mutation co-occurrence associations with the correlation value that is automatically generated by the attention network to discern the underlying relationships between different driver mutations when regulating the immune cytolytic activity (CYT). Using ImmregInformer, we identified 250 immune-regulating driver mutations out of 487 candidates in 8223 pan-cancer samples. The regulatory driver mutations had significantly higher mutation frequency and more cohesive interactions with the cytolytic signature genes, suggesting the accuracy of our identification. Based on the attention module and mutation co-occurrence calculation, we found their complementary roles in exploring the synergistic immune-regulatory mutation pairs. For model interpretability, the CYT-regulatory driver mutations were prioritized in each cancer type based on the weight and regression coefficient. We verified the top-ranked driver mutations exhibited dominant associations with the CYT. In conclusion, this study paves the way for an advanced, interpretable machine learning framework that has potential implications in dissecting the immune regulatory factors in cancer. It also underscores the importance of employing deep learning methods like Transformer to unlock hidden insights into the biological complexities of cancer immunity, offering a new avenue for research on the immune regulatory mechanism and potential clinical applications. ImmregInformer is freely available at https://github.com/Liwen-Liberty/ImmregInformer.

![Image text](https://github.com/Liwen-Liberty/ImmregInformer/blob/main/readmepic/Figure1.png)


To fetch indicators of immune response regulation from tumor somatic mutation profile, we proposed a novel method called ImmregInformer. The workflow was illustrated in Figure 1 for better understanding. At first, the mutation annotation format (MAF) file of somatic mutation was transformed into a binary matrix that indicated the mutation status of all candidate cancer driver genes. This matrix served as input data whose rows corresponded to cancer samples (observations) and columns to the driver mutation status (variables). Each driver mutation variable was then randomly initialized as an embedding vector, and fed to transformer encoder lay-ers. Then, to capture the mutation co-occurrence associations among cancer driver genes, the odd ratio matrix was integrated into the multi-head self-attention module. The dot product of the original mutation status profile and the embeddings generated by the ImmregInformer encoder was calculated to concentrate the information from mutated genes. Subsequently, the modified embeddings encapsulate information regarding the interactions between various driver mutations was served as the final feature vectors. The final feature vector was further utilized in predicting the CYT score through a multi- layer perceptron (MLP) module. The weights in the MLP can be used to interpret the importance of the variables. Since there are challenges in defining an appropriate cut- off for variable weights, the lasso-regularized ordinal re- gression analysis was performed to discern driver mutations with significant immunoregulatory effects. Taking the CYT score as the dependent variable, we utilized the modified embeddings obtained from the Transformer encoder model as the independent variables.
## Table of Contents

- [Installation](#installation)
- [Quick start](#quick-start)
- [Contributing](#contributing)
- [Cite](#cite)
- [Contacts](#contacts)
- [License](#license)


## Installation

scDecipher is tested to work under:

```
* Python 3.8.0
* Torch 1.12.0
* Scanpy 1.9.0
* Anndata 0.8.0
* R 4.2.2
* Numpy 1.23.5
* Other basic python and r toolkits
```
### Installation of other dependencies
* Install [R package glmnetcr](https://github.com/cran/glmnetcr) using ` devtools::install_github("cran/glmnetcr") ` in the R environment if you encounter any issue.


# Quick start
To reproduce our results:

## 1，Application and validation in the identification of immune-regulators
```
cellphonedb method statistical_analysis ./data/RCC_scRNA_P76_metadata.txt ./data/RCC_scRNA_P76_matrix.txt --counts-data=gene_name --threads 100 --output-path ./output/
```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **counts-data** | [ensembl or gene_name or hgnc_symbol] |
| **threads** | Max of threads to process the data. |
| **output-path** | Directory where the results will be allocated (the directory must exist). |

```
Rscript ./tools/run_cellchat.R --count ./data/RCC_scRNA_P76_matrix.txt --meta ./data/RCC_scRNA_P76_metadata.txt  --output ./output/

# The used ligand-target matrix, lr network and weighted networks of interacting cells can be downloaded from [Zenodo](https://zenodo.org/record/7074291).
Rscript ./tools/run_nichenet.R --count ./data/RCC_scRNA_P76_matrix.txt --meta ./data/RCC_scRNA_P76_metadata.txt  --output ./output/

Rscript ./tools/run_icellnet.R --count ./data/RCC_scRNA_P76_matrix.txt --meta ./data/RCC_scRNA_P76_metadata.txt  --output ./output/
```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **count** | Count matrix / normalized count matrix path. |
| **meta** | Meta data (celltypes annotation) path. |
| **output** | Directory where the results will be allocated. |

```
# Obtain the intersection of LR pairs output by 4 cellular communication tools, which are required to be found by at least 2 tools and have expression in scRNA-seq data.
python ./tools/process_final_lr.py --lr_cellphonedb ./output/process_cellphonedb_lr.csv --lr_cellchat ./output/process_cellchat_lr.csv --lr_nichenet ./output/process_nichenet_lr.csv --lr_icellnet ./output/process_icellchat_lr.csv --count ./data/RCC_scRNA_P76_matrix.txt --output ./output/final_lr.csv
```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **lr_cellphonedb** | The results of LR pairs output by cellphonedb. |
| **lr_cellchat** | The results of LR pairs output by cellchat. |
| **lr_nichenet** | The results of LR pairs output by nichenet. |
| **lr_icellnet** | The results of LR pairs output by icellnet. |
| **count** | Count matrix / normalized count matrix path. |
| **output** | The final results of LR pairs. |

<div align="center">
  <img src="https://github.com/Liwen-Liberty/ImmregInformer/blob/main/readmepic/Figure2.png" alt="Editor" width="500">
</div>

===========================================================================

<div align="center">
  <img src="https://github.com/Liwen-Liberty/ImmregInformer/blob/main/readmepic/Figure3.png" alt="Editor" width="500">
</div>

## 2，Interplay of cancer drivers in regulating immune response
```
python ./src/tutorials1/main.py --count ./data/RCC_scRNA_P76_matrix.txt --meta ./data/RCC_scRNA_P76_metadata.txt --lr_file ./output/final_lr.csv --gene CD8A --dca_rank_result ./output/CD8A_dca_rank_result.csv --ccc_ratio_result ./output/CD8A_ccc_ratio_result.csv
```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **count** | Count matrix / normalized count matrix path. |
| **meta** | Meta data (celltypes annotation) path. |
| **lr_file** | The final results of LR pairs. |
| **gene** | The specific target gene name  |
| **dca_rank_result** | The result of prioritize the dominant cell communication assmebly that regulates the target gene expression pattern. |
| **ccc_ratio_result** | The result of ratio of different cell types affected by cellular communication. |

Visualization of results:
<div align="center">
  <img src="https://github.com/Liwen-Liberty/ImmregInformer/blob/main/readmepic/Figure4.png" alt="Editor" width="500">
</div>



## 3，Prioritizing cancer drivers in order of contribution to immune regulation
```
python ./src/tutorials2/main.py --count ./data/RCC_scRNA_P76_matrix.txt --meta ./data/RCC_scRNA_P76_metadata.txt --lr_file ./output/final_lr.csv --gene FOLR2 --cell_type TAM --dca_rank_result ./output/FOLR2_dca_rank_result.csv --ccc_ratio_result ./output/FOLR2_ccc_ratio_result.csv
```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **count** | Count matrix / normalized count matrix path. |
| **meta** | Meta data (celltypes annotation) path. |
| **lr_file** | The final results of LR pairs. |
| **gene** | The specific target gene name.  |
| **cell_type** | The specific cell type (TAM:tumor-associated macrophages).  |
| **dca_rank_result** | The result of prioritize the dominant cell communication assmebly that regulates the target gene expression pattern. |
| **ccc_ratio_result** | The result of ratio of different cell types affected by cellular communication. |



===========================================================================





# Contributing

Jiboya Xuliwen ..

# Cite
<p align="center">
  <a href="https://clustrmaps.com/site/1bpq2">
     <img width="200"  src="https://clustrmaps.com/map_v2.png?cl=ffffff&w=268&t=m&d=4hIDPHzBcvyZcFn8iDMpEM-PyYTzzqGtngzRP7_HkNs" />
   </a>
</p>

<p align="center">
  <a href="#">
     <img src="https://api.visitorbadge.io/api/visitors?path=https%3A%2F%2Fgithub.com%2Fjiboyalab%2FscDecipher&labelColor=%233499cc&countColor=%2370c168" />
   </a>
</p>


# Contacts
If you have any questions or comments, please feel free to email: byj@hnu.edu.cn.

# License

[MIT ? Richard McRichface.](../LICENSE)
